Adding function to `join` multiple `TSFrame`s.

codetalker7 opened the pull request:
This PR addresses issue https://github.com/xKDR/TSFrames.jl/issues/40. We have added a method for `join` which takes in a variable number of `TSFrame`s (this number should be at least 2). We have assumed that the `join`s are left associative. To keep the syntax similar to what we already have, we have added the join type as the keyword argument `jointype` (since the variable number of arguments needs to be the final positional argument). For more, please refer to the discussion in https://github.com/xKDR/TSFrames.jl/issues/40.

Corresponding documentation and tests have been added.

chiraganand commented:
> This PR addresses issue #40. We have added a method for `join` which takes in a variable number of `TSFrame`s (this number should be at least 2). We have assumed that the `join`s are left associative. To keep the syntax similar to what we already have, we have added the join type as the keyword argument `jointype` (since the variable number of arguments needs to be the final positional argument). For more, please refer to the discussion in #40.
> 
> Corresponding documentation and tests have been added.

Yes, this sounds fine. Though, my personal preference would be to reuse the `on` keyword (instead of `jointype`) used by DataFrames.jl to maintain consistency in interfaces, should be easier for users.

codetalker7 commented:
> > This PR addresses issue #40. We have added a method for `join` which takes in a variable number of `TSFrame`s (this number should be at least 2). We have assumed that the `join`s are left associative. To keep the syntax similar to what we already have, we have added the join type as the keyword argument `jointype` (since the variable number of arguments needs to be the final positional argument). For more, please refer to the discussion in #40.
> > Corresponding documentation and tests have been added.
> 
> Yes, this sounds fine. Though, my personal preference would be to reuse the `on` keyword (instead of `jointype`) used by DataFrames.jl to maintain consistency in interfaces, should be easier for users.

Changed `jointype` to `on`.

chiraganand commented:
> Changed `jointype` to `on`.

On second thoughts, `jointype` is a more appropriate name because `on` in DataFrames refers to a column name and not the kind of join, the latter being referred to by the method name itself.

Can you change it back to `jointype`? Sorry about this.

chiraganand reviewed (changes requested):
The documentation also needs to be modified after this API change.

chiraganand commented on src/join.jl:
````diff
@@ -220,6 +232,62 @@ julia> join(ts1, ts2, JoinRight)
+
 ```
 """
 function Base.join(ts1::TSFrame, ts2::TSFrame)
````
The call to `join()` here needs to change to provide the keyword argument name.

chiraganand commented on src/join.jl:
```diff
@@ -247,5 +315,29 @@ function Base.join(ts1::TSFrame, ts2::TSFrame, ::Type{JoinRight})
+    ts1::TSFrame,
+    ts2::TSFrame,
+    ts...;
+    jointype::T
```
`jointype` should take a default value of `JoinAll`.

chiraganand commented on src/join.jl:
```diff
@@ -247,5 +315,29 @@ function Base.join(ts1::TSFrame, ts2::TSFrame, ::Type{JoinRight})
+    if isempty(ts)
+        return Base.join(ts1, ts2, jointype)
+    else
+        return Base.join(Base.join(ts1, ts2, jointype), ts...; jointype=jointype)
```
Will this work with. say, 5 ts objects because the methods above do not accept more than 2 arguments. I thought a better approach could be to make all the methods accept more than 2 args then this method here would directly call the other methods without modifying any arguments.

codetalker7 replied:
Yes, this will work for 5 objects. For example, the following code works.

```julia
using TSFrames, DataFrames, Dates, Random, Statistics
dates1 = collect(Date(2017,1,1):Day(1):Date(2017,1,10))
dates2 = collect(Date(2017,1,1):Day(1):Date(2017,1,30))
ts1 = TSFrame(rand(length(dates1)), dates1)
ts2 = TSFrame(rand(length(dates2)), dates2)
ts3 = TSFrame(rand(length(dates2)), dates2)
ts4 = TSFrame(rand(length(dates2)), dates2)
ts5 = TSFrame(rand(length(dates1)), dates1)

ts_all = join(ts1, ts2, ts3, ts4, ts5)
ts_left = join(ts1, ts2, ts3, ts4, ts5; jointype=JoinLeft)
```
In this method we are just doing the following: if there are exactly 2 ts objects, just call one of the older methods. If there are more than two objects, then compute recursively. This is also how DataFrames.jl implements multiple joins.

We can modify the old methods to have multiple ts objects, but that will just be more code in my opinion. Here we can handle everything with one function.

chiraganand replied:
> We can modify the old methods to have multiple ts objects, but that will just be more code in my opinion. Here we can handle everything with one function.

Wouldn't is be easier to maintain one type of logic in all methods rather than having multiple logics in multiple methods?

chiraganand replied:
Also, the question to ask is whether to even keep the old methods which dispatch on join types? Shouldn't a single method even if it is recursive work? Do we even need the `JoinXXX` types now?

chiraganand replied:
> ... This is also how DataFrames.jl implements multiple joins.

`DataFrames.join*()` functions are recursive then wouldn't `TSFrames.join()` call add another layer of recursive calls on top? Shouldn't the extra layer increase the time complexity by a factor of `log(n)`? In that case, I feel we should collect all the arguments together (`ts1, ts2, ts3, ...`) and then call `DataFrames.join()` in one go.

chiraganand commented on test/join.jl:
```diff
@@ -22,12 +37,28 @@ ts_outerjoin = join(ts1, ts2, JoinOuter)
-@test ts_joinboth[:, :x1] == ts1[1:DATA_SIZE, :x1]
-@test ts_joinboth[1:Int(floor(DATA_SIZE/2)), :x1_1] == ts2[1:Int(floor(DATA_SIZE/2)), :x1]
-@test isequal(Vector{Missing}(ts_joinboth[Int(floor(DATA_SIZE/2)) + 1:DATA_SIZE, :x1_1]), fill(missing, DATA_SIZE - Int(floor(DATA_SIZE/2))))
+ts_joinall = join(ts1, ts2, JoinAll)
```
Can you add another test(s) for testing 4 or 5 ts objects join as a boundary test case as there are already 3 arguments in the method signature so the code might just work for 3?

codetalker7 replied:
Added tests for 5 objects.

codetalker7 commented:
The latest commit replaces everything by a single function. Note that here we are using a `for` loop; we had to do this because some `DataFrames` join function like `leftjoin` and `rightjoin` work on *only* two `DataFrame` objects (and not more than two objects).

chiraganand commented:
> The latest commit replaces everything by a single function. Note that here we are using a `for` loop; we had to do this because some `DataFrames` join function like `leftjoin` and `rightjoin` work on _only_ two `DataFrame` objects (and not more than two objects).

This should be fine as long as there are test cases to make sure the code is working fine for `leftjoin` et al for more than 2 or 3 `TSFrame` objects.

codetalker7 commented:
> > The latest commit replaces everything by a single function. Note that here we are using a `for` loop; we had to do this because some `DataFrames` join function like `leftjoin` and `rightjoin` work on _only_ two `DataFrame` objects (and not more than two objects).
> 
> This should be fine as long as there are test cases to make sure the code is working fine for `leftjoin` et al for more than 2 or 3 `TSFrame` objects.

Hi. Yes, we have tests for 2, 3 and 5 `TSFrame` objects for each function.
